### ISSUE TITLE
Load review completion condition script from repository

### ISSUE THREAD

**Issue by fpeter8** [Reviewable/Reviewable#722]
There are a lot of tools out there that use configuration checked into git itself (travis, Jenkins pipelines, even Github with code owners, pull request template, etc.). I'd like to do the same with reviewable; have a `.js` file under the `.reviewable` folder in root that governs when a PR is complete. This would allow team members without admin access to refine the script and not rely on me to do it for them. (As long as reviewable always uses `master` / whatever custom base branch of the repository this is safe - no one can commit a `return true` in and merge it).

**Comment by pkaminski** [comment]
(I thought there was already an issue open for this but if so I couldn't find it, so I guess this will be it...)

Agreed this would be better in many ways but it turns out to be hard to do under the current authorization model.  Since Reviewable uses GitHub's (legacy) OAuth integration, it needs to impersonate another user for any and all access to a repo.  Unfortunately, it turns out that under Reviewable's current architecture there are situations where we need to access the completion condition but aren't guaranteed to have an appropriate account on hand to do so directly from the repo.  Given that the completion condition is often a key part of the PR approval workflow I don't want to fudge things with caching, skipping evaluation in those cases, etc.

This problem will go away once Reviewable switches to GitHub's App integration.  However, GitHub has been dragging their feet on reaching feature parity, and there's still at least one critical API call that isn't supported in Apps.  If/when this gets resolved I intend to migrate and it should become possible to store the condition in the repo.  (This is being tracked in #482, though admittedly it doesn't have the best title.)

**Comment by jwnimmer-tri** [comment]
As a tip for any other users -- what my org has done to work around this is to version-control the condition on a GitHub Wiki page:

https://github.com/RobotLocomotion/drake-ci/wiki/reviewable.io-review-completion-condition

The GitHub wiki system has rudimentary edit history and diffs available on the web page itself, or you can look at [pull request](https://stackoverflow.com/questions/10642928/how-to-pull-request-a-wiki-page-on-github) tricks if you like.

It does mean manually copying between reviewable and the wiki, but for us changes are infrequent enough that the burden is manageable.

**Comment by pkaminski** [comment]
See #447 for a design that includes the option of loading settings (including the completion condition) from files in a repo.  We don't want to wait for #482, but instead should adopt a copy-in system that listens for pushes on a designated branch in the repo and copies the file contents into Reviewable's database where they can be accessed without credentials (with a reasonably frequent cron job as backup).  If a repo is disconnected it reverts to normal in-page settings, based off the last copied set, since it can no longer guarantee timely updates.

As part of this change we'd also want to offer users a completion condition sandbox similar to the current settings editor, so they could play around with their condition before committing changes.

**Comment by nate-thirdwave** [comment]
Any further planning on this? An alternative could be to just give some basic API access for getting/setting the completion condition blob so we can write a controlled pipeline around updating it.

**Comment by fahhem** [comment]
Hey Nate, we had a false-start on this earlier that we had to rollback. We intend on getting to this after the current (major) Stripe API update though.

**Comment by nate-thirdwave** [comment]
Any further updates on this?  I think at this point, I'd just love a REST or GraphQL API that I can poke at with something like a Terraform script/CI script, etc.

**Comment by nate-thirdwave** [comment]
My core needs are:
 * Change tracking / rollback capabilities.  Right now, if someone updates the condition and breaks something, we don't know what changed or how to go back.   I don't know that Reviewable needs to re-implement git or change history or anything, though it would be nice if the last 4 or 5 versions were stored with a little timeline for visibility.
 * I need to be able to unit-test the completion condition code.   As I see it, it's not Reviewable's job to set that up for me, which is fine, but that means that I need to be able to keep code up to date in both the production (Reviewable) and unit-test contexts.

Overall, I'm happy to effectively "do the work" of managing the history, rollback, unittesting, etc on my side as long as I can set up automated machinery to deploy the code to production.  Perhaps that's a faster pathway than any major UI changes?

**Comment by snoack** [comment]
@nate-thirdwave, I've started working on a mechanism that would allow putting the completion condition (as well as the repository settings) into the repository, or into another repository for global per organization configuration. Would that help your use case?

**Comment by nate-thirdwave** [comment]
That would be fantastic.

**Comment by nate-thirdwave** [comment]
Circling back to this at the 2 month mark. Any progress?

**Comment by snoack** [comment]
It's under development. It's part of a larger change, not only allowing completion conditions to be loaded from the repository, but all settings, along with the option for organization wide master settings.

**Comment by snoack** [comment]
This has been implemented now. For more details refer to the [docs](https://docs.reviewable.io/repositories.html#store-repository-settings-using-the-reviewable-directory).

**Comment by erane-opti** [comment]
@snoack  can you please elaborate a bit on versions/branch behavior? (i.e. only what defined as "default" branch (or main/master)?  if i PR to main, only after merged?, etc..)

**Comment by snoack** [comment]
Which branch is considered the default branch is a setting on GitHub's end. However, for new repositories created through GitHub's UI the default branch is `main` (and was used to be `master` in the past).

Only the contents of the `.reviewable` directory in the repository's default branch (i.e. your `main` or `master` branch if you haven't changed it) are considered by Reviewable.

If a PR targets your default branch, the contents of the PR's source branch are ignored until merged into the default branch.
